### PR TITLE
Add a history task queue manager for Queue V2

### PR DESCRIPTION
**Pull request by MichaelSnowden**
**What changed?**
I added a `history.TaskQueueManager` object which provides some utilities for reading and enqueuing history tasks.

**Why?**
We will need this to implement the history DLQ.

**How did you test it?**
The new logic has 100% test coverage, split across integration tests and unit tests. 

**Potential risks**
This isn't used anywhere yet.

**Is hotfix candidate?**
No.

### CODE REVIEW

File: common/persistence/data_interfaces.go [hunk 1]
```diff
@@ -1204,6 +1205,52 @@ type (
 		SaveClusterMetadata(ctx context.Context, request *SaveClusterMetadataRequest) (bool, error)
 		DeleteClusterMetadata(ctx context.Context, request *DeleteClusterMetadataRequest) error
 	}
+
+	// HistoryTaskQueueManager is responsible for managing a queue of internal history tasks. It is currently unused,
+	// but we plan on using this to implement a DLQ for history tasks. This is called a history task queue manager, but
+	// the actual history task queues are not managed by this object. Instead, this object is responsible for managing
+	// a generic queue of history tasks (which is what the history task DLQ will be).
+	HistoryTaskQueueManager struct {
```
Comment (yycptt) [inline common/persistence/data_interfaces.go:1213]: hmm i guess we probably still need an interface for this. We will need to create metrics client, retryable client, rate limited client, fault injection wrapper for it. 
and do you plan to have a separate PR to adding those impls and wire up with the persistence factory?

File: common/persistence/data_interfaces.go [hunk 2]
```diff
@@ -1204,6 +1205,52 @@ type (
...
+		queue            QueueV2
+		serializer       *serialization.TaskSerializer
+		numHistoryShards int
+	}
+
+	// QueueKey identifies a history task queue. It is converted to a queue name using the GetQueueName method.
+	QueueKey struct {
```
Comment (yycptt) [inline common/persistence/data_interfaces.go:1220]: hmmm, ideally I think this should be an application layer concept and persistence package should depend on the application layer. But in this repo, the relationship is reversed. Guess we have to live with it.

File: common/persistence/history_task_queue_manager.go [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+	//  Raw Task (a proto): <-- when this cannot be deserialized
+	//	- ShardID
+	//	- Blob (a serialized task)
+	ErrDeserializeRawHistoryTask = errors.New("failed to deserialize raw history task from task queue")
+	// ErrDeserializeHistoryTask is returned when the history task cannot be deserialized from the task queue. This
+	// error is returned when the blob inside the raw task cannot be deserialized.
+	//  Raw Task (a proto):
+	//	- ShardID
+	//	- Blob (a serialized task) <-- when this cannot be deserialized
+	ErrDeserializeHistoryTask       = errors.New("failed to deserialize history task blob")
```
Comment (yycptt) [inline common/persistence/history_task_queue_manager.go:60]: should those errors use `SerializationError`, `DeserializationError` or `UnknownEncodingTypeError` type?

File: common/persistence/data_interfaces.go (hunk 1 above)
Comment (MichaelSnowden) [reply common/persistence/data_interfaces.go:1214]: Added a TODO

File: common/persistence/history_task_queue_manager.go (hunk 3 above)
Comment (MichaelSnowden) [reply common/persistence/history_task_queue_manager.go]: Fixed to use the first two.

**Review by yycptt — APPROVED** [review]
